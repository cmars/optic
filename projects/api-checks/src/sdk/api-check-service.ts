import { ApiCheckDsl, Result } from './types';
import { OpenAPIV3 } from 'openapi-types';
import flatten from 'lodash.flatten';
import {
  factsToChangelog,
  OpenAPITraverser,
  IChange,
  IFact,
  OpenApiFact,
} from '@useoptic/openapi-utilities';
import { SpectralDsl } from './spectral/dsl';
import { ApiChangeDsl, ApiCheckDslContext } from './api-change-dsl';

export type DslConstructorInput<Context> = {
  context: Context;
  nextFacts: IFact<OpenApiFact>[];
  currentFacts: IFact<OpenApiFact>[];
  changelog: IChange<OpenApiFact>[];
  nextJsonLike: OpenAPIV3.Document;
  currentJsonLike: OpenAPIV3.Document;
};

export class ApiCheckService<Context> {
  constructor(private getExecutionDate?: (context: Context) => Date) {}

  public rules: ((
    input: DslConstructorInput<Context>
  ) => Promise<Result>[])[] = [];
  public additionalResults: ((
    input: DslConstructorInput<Context>
  ) => Promise<Result[]>)[] = [];

  mergeWith(apiCheckService: ApiCheckService<Context>) {
    this.rules.push(...apiCheckService.rules);
    this.additionalResults.push(...apiCheckService.additionalResults);
  }

  useRulesFrom(rules: (apiChangeDsl: ApiChangeDsl) => void) {
    const dslConstructor = (input: DslConstructorInput<ApiCheckDslContext>) => {
      return new ApiChangeDsl(
        input.nextFacts,
        input.changelog,
        input.currentJsonLike,
        input.nextJsonLike,
        input.context
      );
    };

    const runner = (input: DslConstructorInput<Context>) => {
      const dsl = dslConstructor(input);
      rules(dsl);
      return dsl.checkPromises();
    };

    this.rules.push(runner);
    return this;
  }

  useDsl<DSL extends ApiCheckDsl>(
    dslConstructor: (input: DslConstructorInput<Context>) => DSL,
    ...rules: ((dsl: DSL) => void)[]
  ) {
    const runner = (input: DslConstructorInput<Context>) => {
      const dsl = dslConstructor(input);
      rules.forEach((i) => i(dsl));
      return dsl.checkPromises();
    };

    this.rules.push(runner);
    return this;
  }
  // for our standard DSL
  useRules(rulesMap: { [key: string]: (dsl: ApiChangeDsl) => void }) {
    const dslConstructor = (input: DslConstructorInput<ApiCheckDslContext>) => {
      return new ApiChangeDsl(
        input.nextFacts,
        input.changelog,
        input.currentJsonLike,
        input.nextJsonLike,
        input.context
      );
    };

    const runner = (input: DslConstructorInput<Context>) => {
      const dsl = dslConstructor(input);
      const rules = Object.values(rulesMap);
      rules.forEach((i) => i(dsl));
      return dsl.checkPromises();
    };

    this.rules.push(runner);
    return this;
  }

  // tried using "Ruleset" but getting typeerrors -- falling back to any
  // @Stephen please chech on this
  useSpectralRuleset(ruleset: any) {
    const runner = async (input: DslConstructorInput<Context>) => {
      const dsl = new SpectralDsl(input.nextJsonLike, input.nextFacts, ruleset);
      return await dsl.spectralChecksResults;
    };
    this.additionalResults.push(runner);
    return this;
  }

  useDslWithNamedRules<DSL extends ApiCheckDsl>(
    dslConstructor: (input: DslConstructorInput<Context>) => DSL,
    rulesMap: { [key: string]: (dsl: DSL) => void }
  ) {
    const runner = (input: DslConstructorInput<Context>) => {
      const dsl = dslConstructor(input);
      const rules = Object.values(rulesMap);
      rules.forEach((i) => i(dsl));
      return dsl.checkPromises();
    };

    this.rules.push(runner);
    return this;
  }

  async runRules(
    currentJsonLike: OpenAPIV3.Document,
    nextJsonLike: OpenAPIV3.Document,
    context: Context
  ) {
    const currentTraverser = new OpenAPITraverser();
    const nextTraverser = new OpenAPITraverser();

    await currentTraverser.traverse(currentJsonLike);
    const currentFacts = currentTraverser.accumulator.allFacts();
    await nextTraverser.traverse(nextJsonLike);
    const nextFacts = nextTraverser.accumulator.allFacts();

    const input: DslConstructorInput<Context> = {
      currentFacts,
      nextJsonLike: nextJsonLike,
      currentJsonLike: currentJsonLike,
      nextFacts,
      changelog: factsToChangelog(currentFacts, nextFacts),
      context,
    };

    const checkPromises: Promise<Result>[] = flatten(
      this.rules.map((ruleRunner) => ruleRunner(input))
    );

    const additionalCheckPromises = this.additionalResults.map((ruleRunner) =>
      ruleRunner(input)
    );

    const results: Result[] = await Promise.all(checkPromises);

    const additionalCheckResults: Result[] = flatten(
      await Promise.all(additionalCheckPromises)
    );

    const date = this.getExecutionDate && this.getExecutionDate(context);

    const combinedResults = [...results, ...additionalCheckResults].filter(
      (result) => {
        // filter when effective date is set and context is mapped to a date of execution
        if (result.effectiveOnDate && date) {
          // execution is after effective date, include
          // execution is before effective date filter out
          return date > result.effectiveOnDate;
        }

        return true;
      }
    );

    return combinedResults;
  }
}
