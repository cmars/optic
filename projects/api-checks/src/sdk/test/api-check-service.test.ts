import { ApiCheckService } from '../api-check-service';
import { ExampleDsl, ExampleDslContext } from './example-dsl';
import { OpenAPIV3 } from '@useoptic/openapi-utilities';
export const defaultEmptySpec: OpenAPIV3.Document = {
  openapi: '3.0.1',
  paths: {},
  info: { version: '0.0.0', title: 'Empty' },
};

function completenessApiRules(dsl: ExampleDsl) {
  dsl.operations.changed.must(
    'have consistent operationIds',
    (current, next, context, docs) => {
      const { expect } = require('chai'); // Using Assert style
      docs.includeDocsLink('https://github.com/docs');
      expect(current.operationId).to.equal(
        next.operationId || '',
        'operation ids must be consistent'
      );
    }
  );
  dsl.operations.requirement.must(
    'be able to see spec',
    (value, context, docs, specItem) => {
      expect(specItem).toBeTruthy();
    }
  );
}

it('can run dsl rules through check service', async (done) => {
  const checker = new ApiCheckService<ExampleDslContext>();

  checker.useDsl(
    (input) =>
      new ExampleDsl(input.nextFacts, input.nextJsonLike, input.changelog),
    completenessApiRules
  );

  const results = await checker.runRules(
    {
      ...defaultEmptySpec,
      paths: {
        '/example': {
          get: {
            operationId: 'getExample',
            responses: {},
          },
        },
      },
    },
    {
      ...defaultEmptySpec,
      paths: {
        '/example': {
          get: {
            operationId: 'get_example',
            responses: {},
          },
        },
      },
    },

    { maturity: 'wip' }
  );

  expect(results).toMatchSnapshot();
  done();
});
