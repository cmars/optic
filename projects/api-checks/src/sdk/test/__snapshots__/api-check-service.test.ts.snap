// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`can run dsl rules through check service 1`] = `
Array [
  Object {
    "change": Object {
      "changeType": "changed",
      "changed": Object {
        "after": Object {
          "method": "get",
          "operationId": "get_example",
          "pathPattern": "/example",
        },
        "before": Object {
          "method": "get",
          "operationId": "getExample",
          "pathPattern": "/example",
        },
      },
      "location": Object {
        "conceptualLocation": Object {
          "method": "get",
          "path": "/example",
        },
        "conceptualPath": Array [
          "operations",
          "/example",
          "get",
        ],
        "jsonPath": "/paths/~1example/get",
        "kind": "operation",
      },
    },
    "condition": "have consistent operationIds",
    "docsLink": "https://github.com/docs",
    "effectiveOnDate": undefined,
    "error": "operation ids must be consistent: expected 'getExample' to equal 'get_example'",
    "isMust": true,
    "isShould": false,
    "passed": false,
    "where": "updated operation: GET /example",
  },
  Object {
    "change": Object {
      "location": Object {
        "conceptualLocation": Object {
          "method": "get",
          "path": "/example",
        },
        "conceptualPath": Array [
          "operations",
          "/example",
          "get",
        ],
        "jsonPath": "/paths/~1example/get",
        "kind": "operation",
      },
      "value": Object {
        "method": "get",
        "operationId": "get_example",
        "pathPattern": "/example",
      },
    },
    "condition": "be able to see spec",
    "docsLink": undefined,
    "effectiveOnDate": undefined,
    "isMust": true,
    "isShould": false,
    "passed": true,
    "where": "requirement for operation: GET /example",
  },
]
`;
